pandas/core/groupby/grouper.py

```diff
@@ -348,9 +349,19 @@ def _set_grouper(self, obj: FrameOrSeries, sort: bool = False):
             if getattr(self.grouper, "name", None) == key and isinstance(
                 obj, ABCSeries
             ):
-                # pandas\core\groupby\grouper.py:348: error: Item "None" of
-                # "Optional[Any]" has no attribute "take"  [union-attr]
-                ax = self._grouper.take(obj.index)  # type: ignore[union-attr]
+                # Sometimes (when self.indexer is not None) self._grouper will be
+                # sorted while obj is not. In this case there is a mismatch when we
+                # call self._grouper.take(obj.index) so we need to undo the sorting
+                # before we call _grouper.take.
+                if self.indexer is not None:
+                    assert self._indexer is not None
+                    assert self._grouper is not None
+                    reverse_indexer = self._indexer.argsort()
+                    unsorted_ax = self._grouper.take(reverse_indexer)
```

Comment:
why are you setting `self._indexer` itself?  this is *only* defined above (and not in the Grouper at all).

Reply:
if anything you should be setting `self.indexer`

Reply:
This is kind of complicated and not so easy to explain without actually showing you the code running through a debugger but it's for the exact same reason that self._grouper is set which is something that was already in the code before me.  I'll try to refer to the test I wrote to make it a little more concrete.  

https://github.com/pandas-dev/pandas/blob/34e00be56ac0580bfcd18111858c447c16d3e21c/pandas/tests/resample/test_resampler_grouper.py#L396

The self.grouper object will initially become a DatetimeIndex which contains all 8 of the dates from the date column in the dataframe.  But then later as the code gets to the lines in question its in order to compute things specifically for cat_1 and cat2.  When it computes things for cat_1 the self.grouper object would get changed to being only the 5 entries from cat_1 and the rest of the 8 dates would be lost if they weren't saved in self._grouper instead.  It's the same issue with self.indexer.  It ends up being reset below in line 388 to correspond to only a subset of the dataframe that belongs to cat_1 so if I don't save it as self._indexer the rest of the information will be lost.

Let me know if you need more explanation on this.  I know my explanation is kind of convoluted but it's not easy to make it clear without actually showing you the code running and being able to answer your questions in real time.

Reply:
can you see if you can find a better condition, this whole _group,er, grouper bizness (which is my fault originally :-<) is kind of smelly and would like to avoid repeating this.

I think you might be able to add a property to a grouper to figure this out. 

something is mutating grouper state.

i like the remedy, but looking to make it more understand / less complex.

Reply:
I've added grouper_resorted attribute on the TimeGrouper that defaults to False upon creation.  When TimeGrouper.grouper is sorted I then set that attribute to True and now I'm using that as the condition to check whether to undo the sorting rather checking if self.indexer is None.  I do still have to save self._indexer otherwise the information on how to undo the sorting will be lost but I think the new condition makes the code a little more understandable.  Let me know if this was what you had in mind.

Reply:
I also set self._indexer = None right below where self._grouper = None is set in Grouper.__init__ to make things a little more transparent.